### ISSUE TITLE
Avoid duplicate error messages due to cross-product

### ISSUE THREAD

**Issue by shaobo-he-aws** [cedar-policy/cedar#248]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/cedar-policy/cedar/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/cedar-policy/cedar/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### Bug Category

Schemas and Validation

### Describe the bug

The validator can produce the same error message multiple times, one for each product configuration.

### Expected behavior

It should remove the duplicates. Alternatively, make the error more specific in each case, so they are not duplicates.

### Reproduction steps

cedar validate -s example.cedarschema.json -p example.cedar

### Code Snippet

Example policy,
```
permit (
  principal,
  action in [Action::"SomeAction"],
  resource
)
when { principal in resource.someattribute };
```

Example schema,
```JSON
{
  "My::Name::Space": {
    "entityTypes": {
      "Entity1": {
      },
      "Entity2": {
        "memberOfTypes": ["Entity3"],
        "shape": {
          "type": "Record",
          "attributes": {
            "attr4": {
              "type": "Entity",
              "name": "Entity4"
            }
          }
        }
      },
      "Entity3": {
      },
      "Entity4": {
        "memberOfTypes": [],
        "shape": {
          "type": "Record",
          "attributes": {
            "attr1": {
              "type": "Entity",
              "name": "Entity1"
            }
          }
        }
      }
    },
    "actions": {
      "action2": {
        "appliesTo": {
          "principalTypes": ["Entity1"],
          "resourceTypes": ["Entity2"]
        }
      },
      "action4": {
        "appliesTo": {
          "principalTypes": ["Entity1"],
          "resourceTypes": ["Entity4"]
        }
      },
      "action3": {
        "appliesTo": {
          "principalTypes": ["Entity1"],
          "resourceTypes": ["Entity3"]
        }
      }
    }
  }
}
```

### Log output

```
Validation Results:
Validation error on policy policy0: Unrecognized action id Action::"SomeAction", did you mean My::Name::Space::Action::"action2"?
Validation error on policy policy0: Unable to find an applicable action given the policy head constraints
Validation error on policy policy0 at offset 112-134: Attribute not found in record or entity: someattribute
Validation error on policy policy0 at offset 112-134: Attribute not found in record or entity: someattribute
Validation error on policy policy0 at offset 112-134: Attribute not found in record or entity: someattribute
```

**Comment by andrewmwells-amazon** [comment]
Personally I think we should include the cross product with every error (https://github.com/cedar-policy/cedar/issues/135). In that case, I don't think it's that bad to have the duplicates.

**Comment by john-h-kastner-aws** [comment]
This is a bug in how we forward the error messages. The validator collects errors messages in a hash map, so exact duplicates are already avoided. The three errors returned by the validator are actually not the same even though the text is identical. The validator returns a "suggested fix" with each that is not being printed. The resource entity types all have different attributes, so the suggestion is different.

**Comment by john-h-kastner-aws** [comment]
Fix the immediate issue in #252. It will still display three very similar errors messages. As future work, we could do something clever to gather error messages which have the same root cause but different details into a single message which would explain the error once and then elaborate on the individual details.

**Comment by john-h-kastner-aws** [comment]
Closing as fixed by #252.
